SpiceSharp/Algebra/Solve/ComplexSolver.cs

```diff
@@ -0,0 +1,333 @@
+
+            // Setup for reordering
+            Strategy.Setup(Matrix, Rhs, step, Magnitude);
+
+            // Perform reordering and factorization starting from where we stopped last time
+            for (; step <= Matrix.Size; step++)
+            {
+                var pivot = Strategy.FindPivot(Matrix, step);
+                if (pivot == null)
+                    throw new SparseException("Singular matrix");
+
+                // Move the pivot to the current diagonal
+                MovePivot(pivot, step);
+
+                // Elimination
+                Elimination(pivot);
+            }
+        }
+
+        /// <summary>
+        /// Eliminate a row
+        /// </summary>
+        /// <param name="pivot">Current pivot</param>
+        void Elimination(MatrixElement<Complex> pivot)
+        {
+            // Test for zero pivot
+            if (pivot.Value.Equals(0.0))
+                throw new SparseException("Matrix is singular");
+            pivot.Value = Inverse(pivot.Value);
+
+            var upper = pivot.Right;
+            while (upper != null)
+            {
+                // Calculate upper triangular element
+                // upper = upper / pivot
+                upper.Value *= pivot.Value;
+
+                var sub = upper.Below;
+                var lower = pivot.Below;
+                while (lower != null)
+                {
+                    int row = lower.Row;
+
+                    // Find element in row that lines up with the current lower triangular element
+                    while (sub != null && sub.Row < row)
+                        sub = sub.Below;
+
+                    // Test to see if the desired element was not found, if not, create fill-in
+                    if (sub == null || sub.Row > row)
+                        sub = base.CreateFillin(row, upper.Column);
+
+                    // element -= upper * lower
+                    sub.Value -= upper.Value * lower.Value;
+                    sub = sub.Below;
+                    lower = lower.Below;
+                }
+                upper = upper.Right;
+            }
+        }
+
+        /// <summary>
+        /// Magnitude method
+        /// </summary>
+        /// <param name="value">Value</param>
+        /// <returns></returns>
+        static double Magnitude(Complex value) => Math.Abs(value.Real) + Math.Abs(value.Imaginary);
+
+        /// <summary>
+        /// Calculate inverse of a complex number
+        /// </summary>
+        /// <param name="value">Value</param>
+        /// <returns></returns>
+        static Complex Inverse(Complex value)
+        {
+            double real, imaginary;
+            double r;
+            if ((value.Real >= value.Imaginary && value.Real > -value.Imaginary) ||
+                (value.Real < value.Imaginary && value.Real <= -value.Imaginary))
+            {
+                r = value.Imaginary / value.Real;
+                real = 1.0 / (value.Real + r * value.Imaginary);
```

Comment:
@svenboulanger Can `value.Real + r * value.Imaginary`  or value.Real  be `0`?

Reply:
Basically, the conditional statements will try to divide by the maximum of value.Real and value.Imaginary to reduce numerical errors.
If both are 0, then the result will/should be NaN.